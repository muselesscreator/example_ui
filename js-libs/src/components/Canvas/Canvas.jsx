/** @module */

import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Component from '../Component';
import { shouldUpdate } from 'modules/react';
import CanvasAPI from './api';

import './Canvas.scss';

/**
 * @callback mouseEventCallback
 * @param {Number[]} coordinates - event coordinates.
 */

/**
 * @callback zoomCallback
 * @param {Object} transform - new transform
 */

/**
 * Canvas component wraps a simple canvas and provides a number of
 * access features to its context, while also attaching a zoom/pan
 * listener.
 * 
 * When attaching a ref to this component, you gain direct access to:
 * - this.canvas - ref to internal canvas
 * - this.d3Canvas - d3 reference to canvas
 * - this.ctx - canvas context
 * - this.fitToScreen() - resize content back to canvas size.
 *
 * @param {Number} height - canvas height (pixels)
 * @param {Number} width - canvas width (pixels)
 * @param {mouseEventCallback} onmousemove - function to be called with
 *   the mouse position of mousemove events.
 * @param {mouseEventCallback} onmouseout - function to be called on mouseout events.
 * @param {zoomCallback} onzoom - function to be called with new transform on drag/zoom
 */
export class Canvas extends Component {
  constructor(props) {
    super(props);
    this.state = {
      transform: { x: 0, y: 0, k: 1 },
    }
    this.canvas = new CanvasAPI(this);
  }
  shouldComponentUpdate(prevProps, prevState) {
    return shouldUpdate(this, prevProps, prevState);
  }
  componentDidMount() {
    this.canvas.initialize();
  }
  render() {
    const { className, height, width } = this.props;
    return (
      <div
        className={classNames('brw-canvas', className)}
        style={{width: `${width}px`, height: `${height}px`}}
      >
        <canvas ref="canvas" width={width} height={height} />
      </div>
    );
  }
}

Canvas.propTypes = {
  height: PropTypes.number,
  onmousemove: PropTypes.func,
  onmouseout: PropTypes.func,
  onZoom: PropTypes.func,
  width: PropTypes.number,
};
Canvas.defaultProps = {
  onZoom: () => ({}),
  onmousemove: () => ({}),
  onmouseout: () => ({}),
}

export default Canvas;
