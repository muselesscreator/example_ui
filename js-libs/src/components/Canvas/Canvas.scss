.brw-canvas {
  position: relative;
  canvas {
    position: absolute;
    top: 0px;
    left: 50%;
    transform: translate(-50%, 0);
    background-color: black;
    &.cover {
      visibility: hidden;
    }
  }
}
