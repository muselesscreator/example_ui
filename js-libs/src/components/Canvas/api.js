import * as d3 from 'd3';

/**
 * Canvas component wraps a simple canvas and provides a number of
 * access features to its context, while also attaching a zoom/pan
 * listener.
 * 
 * When attaching a ref to this component, you gain direct access to:
 * * this.canvas - ref to internal canvas
 * * this.d3Canvas - d3 reference to canvas
 * * this.ctx - canvas context
 * * this.fitToScreen() - resize content back to canvas size.
 *
 * @param height {number} - canvas height (pixels)
 * @param width {number} - canvas width (pixels)
 * @param draw {func} - function to be called on re-render (drag/zoom)
 * @param onmousemove {func} - function to be called with the mouse position of
 *                             mousemove events.
 * @param onmouseout {func} - function to be called on mouseout events.
 * @param onzoom {func} - function to be called with new transform on drag/zoom
 */
export class Canvas extends Object {
  constructor(el) {
    super(el.props);
    this.el = el;
    this.gridZoom = null;
    this.initialScale = 3;
    this.scaleExtent = [1/8, 20];
    this.transform = { x: 0, y: 0, k: 1 };
    this.imgProps = [ null, 0, 0, 1, 1 ];
  }

  initialize() {
    this.enableZoom();
    this.draw();
    const d3Canvas = this.d3Canvas();
    d3Canvas.on("mousemove", this.onmousemove);
    d3Canvas.on("mouseout", this.onmouseout);
  }

  /**
   * @return { object } - element props
   */
  props = () => this.el.props;

  /**
   * Calls the element's setState method with the passed args.
   */
  setState = (...args) => this.el.setState(...args);

  /**
   * @return - canvas element
   */
  canvas = () => this.el.refs.canvas;

  /**
   * @return - d3 canvas selection
   */
  d3Canvas = () => d3.select(this.canvas());

  /**
   * @return - canvas context
   */
  ctx = () => this.canvas().getContext('2d');

  /**
   * Callback for onmousemove events.
   *
   * Calls passed onmousemove callback from element with scaled location.
   */
  onmousemove = () => {
    const { offsetX, offsetY } = d3.event;
    const { x, y, k } = this.transform;
    this.props().onmousemove([
      (offsetX - x) / k,
      (offsetY - y) / k,
    ]);
  }

  /**
   * Callback for onmouseout events.
   *
   * Calls passed onmouseout callback from element.
   */
  onmouseout = () => {
    this.props().onmouseout();
  }

  /**
   * Re-draws last image.
   */
  draw = () => {
    if (this.imgProps[0] !== null) {
      this.drawImage(...this.imgProps);
    }
  }

  /**
   * Calls drawImage on the canvas context with the passed args
   * after setting the translation and scal eot match the trasnform.
   */
  drawImage(...props) {
    const ctx = this.ctx();

    this.imgProps = props;

    ctx.save();

    ctx.translate(this.transform.x, this.transform.y);
    ctx.scale(this.transform.k, this.transform.k);
    ctx.drawImage(...this.imgProps);

    ctx.restore();
  }

  /**
   * Resets the canvas by toggling its width.
   */
  clear = () => this.canvas().width = this.props().width;

  /**
   * Apply zoom transform on zoom event.
   */
  zoomed = (transform) => {
    this.props().onZoom(transform);
    this.transform = transform;
    this.clear();
    this.draw();
  }

  /**
   * Register zoom interaction callback.
   */
  enableZoom = () => {
    const zoomed = () => this.zoomed(d3.event.transform);
    const { width, height } = this.props();

    const w = this.initialScale * width;
    const h = this.initialScale * height;

    this.gridZoom = d3.zoom()
      .scaleExtent(this.scaleExtent)
      .translateExtent([[-w, -h], [w, h]])
      .on("zoom", zoomed);
    this.d3Canvas().call(this.gridZoom);
  }

  /**
   * Apply identity transform
   */
  fitToScreen = () => {
    this.zoomed({x: 0, y: 0, k: 1});
  }
}

export default Canvas;
