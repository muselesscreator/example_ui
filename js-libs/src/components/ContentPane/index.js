import ContentPane from './ContentPane';

export default ContentPane;
