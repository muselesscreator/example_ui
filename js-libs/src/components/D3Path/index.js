import D3Path from './D3Path';

export default D3Path;
