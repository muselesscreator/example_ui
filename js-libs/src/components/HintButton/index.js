import HintButton from './HintButton';

export default HintButton;
