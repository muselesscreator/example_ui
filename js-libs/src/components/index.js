export { default as Accordion } from './Accordion';
export { default as Button } from './Button';
export { default as Canvas } from './Canvas';
export { default as Checkbox } from './Checkbox';
export { default as Component } from './Component';
export { default as ContentColumn } from './ContentColumn';
export { default as ContentPane } from './ContentPane';
export { default as D3Path } from './D3Path';
export { default as Divider } from './Divider';
export { default as Dropzone } from './Dropzone';
export { default as FormErrors } from './FormErrors';
export { default as HintButton } from './HintButton';
export { default as HintControl } from './HintControl';
export { default as HintLabel } from './HintLabel';
export { default as Input } from './Input';
export { default as MenuButton } from './MenuButton';
export { default as ModalButton } from './ModalButton';
export { default as Nav } from './Nav';
export { default as Number } from './Number';
export { default as Select } from './Select';
export { default as Slider } from './Slider';
export { default as Table } from './Table';
export { default as ValueLabel } from './ValueLabel';
