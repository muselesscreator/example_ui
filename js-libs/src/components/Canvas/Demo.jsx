import React from 'react';

import Component from '../Component';
import Canvas from './Canvas';

export class Demo extends Component {
  constructor(props) {
    super(props);
    this.canvas = null;
  }
  componentDidMount() {
    this.canvas = this.refs.canvas;
  }
  draw = (ctx) => {
    ctx.rect(20, 20, 150, 150);
    ctx.strokeStyle="#FFFFFF";
    ctx.stroke();
  }
  onZoom = (transform) => {
    console.log(transform);
  }
  render() {
    return (
      <Canvas
        ref="canvas"
        width={300}
        height={300}
        draw={this.draw}
        onZoom={this.onZoom}
      />
    );
  }
}

export default Demo;
