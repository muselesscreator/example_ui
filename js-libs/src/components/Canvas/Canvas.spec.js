import React from 'react';

import { mount } from 'enzyme';

import { Canvas } from './Canvas';
import { simpleTest } from 'modules/redux-test/components';

import CanvasAPI from './api';

jest.mock('./api', () => {
  return jest.fn().mockImplementation((args) => ({
    args,
    initialize: jest.fn(),
  }))
});

describe('Canvas component', () => {
  simpleTest('smoke', Canvas, {
    height: 20,
    width: 100,
    className: 'cl_n',
    onZoom: jest.fn(),
    onmousemove: jest.fn(),
    ommouseout: jest.fn(),
  });
  it('includes default functions for event props', () => {
    const el = mount(<Canvas />);
    expect(el.prop('onZoom')()).toEqual({});
    expect(el.prop('onmousemove')()).toEqual({});
    expect(el.prop('onmousemove')()).toEqual({});
  });
  it('initializes a Canvas API module with itself', () => {
    const props = { some: 'props', for_a: 'component test' };
    const el = mount(<Canvas { ...props } />);
    expect(el.state()).toEqual({ transform: { x: 0, y: 0, k: 1 } });
    const instance = el.instance();
    expect(instance.canvas.args).toEqual(instance);
    expect(instance.canvas.initialize).toHaveBeenCalled();
  });
});
