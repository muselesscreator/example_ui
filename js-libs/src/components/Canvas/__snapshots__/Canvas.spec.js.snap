// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas component :smoke 1`] = `
<div
  className="brw-canvas cl_n"
  style={
    Object {
      "height": "20px",
      "width": "100px",
    }
  }
>
  <canvas
    height={20}
    width={100}
  />
</div>
`;
